src/main/java/com/google/api/codegen/csharp/CSharpDiscoveryContext.java

```diff
@@ -0,0 +1,232 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.csharp;
+
+import com.google.api.Service;
+import com.google.api.codegen.ApiaryConfig;
+import com.google.api.codegen.DiscoveryContext;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Api;
+import com.google.protobuf.Field;
+import com.google.protobuf.Method;
+import com.google.protobuf.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+public class CSharpDiscoveryContext extends DiscoveryContext implements CSharpContext {
+
+  private static final ImmutableMap<Field.Kind, String> FIELD_TYPE_MAP =
+      ImmutableMap.<Field.Kind, String>builder()
+          .put(Field.Kind.TYPE_UNKNOWN, "object")
+          .put(Field.Kind.TYPE_BOOL, "bool")
+          .put(Field.Kind.TYPE_INT32, "int")
+          .put(Field.Kind.TYPE_UINT32, "uint")
+          .put(Field.Kind.TYPE_INT64, "long")
+          .put(Field.Kind.TYPE_UINT64, "ulong")
+          .put(Field.Kind.TYPE_FLOAT, "float")
+          .put(Field.Kind.TYPE_DOUBLE, "double")
+          .put(Field.Kind.TYPE_STRING, "string")
+          .build();
+
+  private static final ImmutableMap<Field.Kind, String> DEFAULTVALUE_MAP =
+      ImmutableMap.<Field.Kind, String>builder()
+          .put(Field.Kind.TYPE_UNKNOWN, "null")
+          .put(Field.Kind.TYPE_BOOL, "false")
+          .put(Field.Kind.TYPE_INT32, "0")
+          .put(Field.Kind.TYPE_UINT32, "0U")
+          .put(Field.Kind.TYPE_INT64, "0L")
+          .put(Field.Kind.TYPE_UINT64, "0UL")
+          .put(Field.Kind.TYPE_FLOAT, "0.0f")
+          .put(Field.Kind.TYPE_DOUBLE, "0.0")
+          .put(Field.Kind.TYPE_STRING, "\"\"")
+          .build();
+
+  private CSharpContextCommon csharpCommon;
+
+  protected CSharpDiscoveryContext(Service service, ApiaryConfig apiaryConfig) {
+    super(service,apiaryConfig);
+  }
+
+  @Override
+  public void resetState(CSharpContextCommon csharpCommon) {
+    this.csharpCommon = csharpCommon;
+  }
+
+  public String addImport(String namespace) {
+    return csharpCommon.addImport(namespace);
+  }
+  
+  private String packageNameAndImport() {
+    String packageName = CSharpContextCommon.s_underscoresToPascalCase(
+        getApiaryConfig().getServiceCanonicalName());
+    String versionName = getApiaryConfig().getServiceVersion().replace('.', '_');
```

Comment:
Yes, I've just checked, and most are major version only.
A counter-example is prediction, which is currently v1.6, but the namespace is Google.Apis.Prediction.v1_6. This code is to match that namespace when minor version is included.